### PR TITLE
Add otlp log extension methods for LoggerProviderBuilder

### PR DESCRIPTION
**Pull request by Yun-Ting**
Towards https://github.com/open-telemetry/opentelemetry-dotnet/issues/4433
Based on https://github.com/open-telemetry/opentelemetry-dotnet/pull/4596

## Changes

* Added support for OTEL1000 Experimental-APIs:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/diagnostics/experimental-apis/OTEL1000.md.
* Added support in OTLP Exporter to bring feature parity (with metrics and tracing) to logging
under experimental flag for dependency Injection.
* Resolved circular reference issue if some service used inside `LoggerProvider` needs access to `ILoggerFactory` (for example an exporter trying to use `IHttpClientFactory` to construct an `HttpClient` instance)

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 1]
```diff
@@ -27,6 +27,139 @@ public class OtlpLogExporterTests : Http2UnencryptedSupportTests
...
+        {
+            var factory = serviceProvider.GetRequiredService<ILoggerFactory>();
+            Assert.NotNull(factory);
+        }
+    }
+
+    [Fact]
+    public void AddOtlpExporterWithNamedOptions()
+    {
+        int defaultExporterOptionsConfigureOptionsInvocations = 0;
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:56]:
```suggestion
        int defaultConfigureExporterOptionsInvocations = 0;
```

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 2]
```diff
@@ -27,6 +27,139 @@ public class OtlpLogExporterTests : Http2UnencryptedSupportTests
...
+        int namedExporterOptionsConfigureOptionsInvocations = 0;
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:57]:
```suggestion
        int namedConfigureExporterOptionsInvocations = 0;
```

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 3]
```diff
@@ -27,6 +27,139 @@ public class OtlpLogExporterTests : Http2UnencryptedSupportTests
...
+            Assert.Equal(2, invocations);
+        }
+
+        options.HttpClientFactory = null;
+        Assert.Throws<InvalidOperationException>(() =>
+        {
+            using var exporter = new OtlpLogExporter(options);
+        });
+
+        options.HttpClientFactory = () => null;
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:117]: why is this twice?

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 4]
```diff
@@ -27,6 +27,139 @@ public class OtlpLogExporterTests : Http2UnencryptedSupportTests
 {
     private static readonly SdkLimitOptions DefaultSdkLimitOptions = new();
 
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public void ResolutionOrderTest(bool requestLoggerProviderDirectly)
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:33]: @Yun-Ting \ @CodeBlanch  - Could we add a short explanation on why we are adding this test?

File: src/OpenTelemetry.Api.ProviderBuilderExtensions/AssemblyInfo.cs [hunk 5]
```diff
@@ -6,8 +6,10 @@
 [assembly: InternalsVisibleTo("OpenTelemetry" + AssemblyInfo.PublicKey)]
 [assembly: InternalsVisibleTo("OpenTelemetry.Tests" + AssemblyInfo.PublicKey)]
 [assembly: InternalsVisibleTo("OpenTelemetry.Api.ProviderBuilderExtensions.Tests" + AssemblyInfo.PublicKey)]
+[assembly: InternalsVisibleTo("OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests" + AssemblyInfo.PublicKey)]
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Api.ProviderBuilderExtensions/AssemblyInfo.cs:9]: Should this moved inside `#if !EXPOSE_EXPERIMENTAL_FEATURES`?

Comment (CodeBlanch) [reply src/OpenTelemetry.Api.ProviderBuilderExtensions/AssemblyInfo.cs]: Updated

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 4 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: @vishweshbankwar I removed the test from here since it really didn't have anything to do with OtlpExporter and replaced it with one in `OpenTelemetry.Tests` called `CircularReferenceTest` which should be more self-explanatory. LMK!

Comment (CodeBlanch) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: Diff for easy reviewing: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5103/commits/e208fbcf8ece49f4f96fc2df58f334c21cc4739b

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 3 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: I took a look, doesn't seem to be a reason for it. Probably copy/paste fail. I removed the extra block.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 2 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: Updated

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 1 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: Updated names and also added `ExperimentalOptions` so we have coverage there.

**Comment by CodeBlanch** [comment]
@utpilla @vishweshbankwar I'm not going to approve this because I did some of the work and I don't want to self-approve but just FYI I think it is in a good state for final review/merge.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs [hunk 6]
```diff
@@ -106,29 +109,280 @@ public static OpenTelemetryLoggerOptions AddOtlpExporter(this OpenTelemetryLogge
...
+    {
+        var finalOptionsName = name ?? Options.DefaultName;
+
+        builder.ConfigureServices(services =>
+        {
+            if (name != null && configureExporter != null)
+            {
+                // If we are using named options we register the
+                // configuration delegate into options pipeline.
+                services.Configure(finalOptionsName, configureExporter);
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs:214]: For the other AddOtlpExporter method which takes in an `Action<OtlpExporterOptions, LogRecordExportProcessorOptions>`, we call the configure action on the options in the `LoggerProviderBuilder.AddProcessor` method regardless of whether the name is null or not.

Could we be consistent with that? Is there any benefit to configuring options differently in this method?

File: src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggingExtensions.cs [hunk 7]
```diff
@@ -206,10 +213,41 @@ public static class OpenTelemetryLoggingExtensions
...
+                        /*
+                         * Note:
+                         *
+                         * There is a possibility of a circular reference when
+                         * accessing LoggerProvider from the IServiceProvider.
+                         *
+                         * If LoggerProvider is the first thing accessed, and
+                         * it requires some service which accesses ILogger
+                         * (for example, IHttpClientFactory), then the
+                         * OpenTelemetryLoggerProvider will try to access the
```
Comment (utpilla) [inline src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggingExtensions.cs:232]: >  OpenTelemetryLoggerProvider will try to access the LoggerProvider inside the initial access to LoggerProvider.

"access LoggerProvider inside the initial access to LoggerProvider". Is there some typo here or is it just me finding it hard to follow? 😄

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs [hunk 8]
```diff
@@ -106,29 +109,280 @@ public static OpenTelemetryLoggerOptions AddOtlpExporter(this OpenTelemetryLogge
...
-        }
-
-        Debug.Assert(exporterOptions != null, "exporterOptions was null");
-        Debug.Assert(processorOptions != null, "processorOptions was null");
+        Debug.Assert(sp != null, "sp was null");
 
-        var config = sp.GetRequiredService<IConfiguration>();
+        var config = sp!.GetRequiredService<IConfiguration>();
 
         var sdkLimitOptions = new SdkLimitOptions(config);
```
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs:344]: The newly added `AddOtlpLogExporter` methods are registering OptionsFactory for `SdkLimitOptions` and `ExperimentalOptions`.

Shouldn't we be using `IOptionsFactory<TOptions>.Create` method to create these options?

File: test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs [hunk 9]
```diff
@@ -266,7 +266,54 @@ public void VerifyExceptionIsThrownWhenImplementationFactoryIsNull()
...
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public void CircularReferenceTest(bool requestLoggerProviderDirectly)
+    {
+        var services = new ServiceCollection();
+
+        services.AddLogging(logging => logging.AddOpenTelemetry());
+
+        services.ConfigureOpenTelemetryLoggerProvider(builder => builder.AddProcessor<TestLogProcessorWithILoggerFactoryDependency>());
```
Comment (utpilla) [inline test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs:278]: Could we specifically test to check that this processor is actually added to the SDK by inspecting the `LoggerProviderSdk` instance that was registered in the service collection?

Comment (Yun-Ting) [reply test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs:278]: The above mentioning check was already covered in this [test](https://github.com/open-telemetry/opentelemetry-dotnet/blob/66ea5e39a620eea2d9f492a8d5eccb62d187614b/test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs#L247C1-L249C67).

The CircularReferenceTest is specifically to cover a scenario that will be supported in the future.
See: [scenario](https://github.com/open-telemetry/opentelemetry-dotnet/blob/66ea5e39a620eea2d9f492a8d5eccb62d187614b/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs#L366C1-L380C12) and [circular reference test comment](https://github.com/open-telemetry/opentelemetry-dotnet/blob/66ea5e39a620eea2d9f492a8d5eccb62d187614b/test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs#L304C1-L308C46)

File: src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggingExtensions.cs (hunk 7 above)
Comment (Yun-Ting) [reply src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggingExtensions.cs:232]: I think @CodeBlanch wrote this big block of comment trying to explain the below scenario.
And I am not sure how to make this comment easier to follow now. 


```mermaid
flowchart TD
    A[ServiceCollection] -->|access| B(LoggerProvider)
    B --> C(HttpClientFactory)
    C --> D(ILogger)
    D --> B

E[LoggerProviderBuilderSdk] --> |use State.Provider to check whether it has already been registered| A[ServiceCollection]
```

File: test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs (hunk 9 above)
Comment (utpilla) [reply test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs:278]: That [test](https://github.com/open-telemetry/opentelemetry-dotnet/blob/66ea5e39a620eea2d9f492a8d5eccb62d187614b/test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs#L247C1-L249C67) is testing against a regular processor.

This circular reference unit test should test if the `LoggerProviderSdk` can actually add a processor that causes a circular reference.

File: src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggingExtensions.cs (hunk 7 above)
Comment (utpilla) [reply src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggingExtensions.cs:232]: Probably rephrase it to something like this, if the intention is to reuse the already created `LoggerProvider` instead of creating a new one:

> If LoggerProvider is the first thing accessed, and it requires some service which accesses ILogger (for example, IHttpClientFactory), then a `LoggerProvider` instance would have already been created and we should use that instance for creating `OpenTelemetryLoggerProvider`.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs (hunk 8 above)
Comment (Yun-Ting) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs:344]: The `SdkLimitOptions` and `ExperimentalOptions` will be registered into the `IOptionsFactory<T>` which takes IConfiguration as one of the parameters:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/e3759a1e0ea6e23b80cad33a206795fda8e240ff/src/Shared/Options/ConfigurationExtensions.cs#L137C1-L145C12
So the options can be retrieved when constructing new option instances with `IConfiguration`.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs (hunk 6 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs:214]: This is why I don't like the pattern where we expose a delegate which takes two parameters (ex `Action<OtlpExporterOptions, LogRecordExportProcessorOptions>`). There's no way to do that correctly with Options API. You can only execute it inline which creates some rough edges for users who expect "Options API things" to work. For example there is the `IPostConfigureOptions` thing where you can register some delegate to fire after other configurations. The inline call won't play nice with that. It will always be after all the configure calls and the post configure calls. Essentially it is inconsistent because we try to support Options API fully where we can.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs (hunk 8 above)
Comment (utpilla) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs:344]: My question is instead of using the regular ctor for `SdkLimitOptions` and `ExperimentalOptions`, shouldn't we be using either: `IOptionsFactory<TOptions>.Create` or `IOptionsMonitor<TOptions>.Get` to get these options. Shouldn't we be utilizing the `OptionsFactory` that we are now registering in this PR for these two options?

File: src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggingExtensions.cs (hunk 7 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggingExtensions.cs]: I pushed a revision. It may or may not help the situation 🤣

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs (hunk 6 above)
Comment (utpilla) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs:214]: Thanks for the explanation!

File: test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs (hunk 9 above)
Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/Logs/OpenTelemetryLoggingExtensionsTests.cs:278]: I do think it is kind of overkill and we have other tests validating whatever you register shows up but no big deal to add an extra check, updated.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs (hunk 8 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpLogExporterHelperExtensions.cs]: It isn't as important to use Options API with `SdkLimitOptions` and `ExperimentalOptions` because they are `internal` but I went ahead and refactored it so it is nice and consistent now.

**Review by utpilla — APPROVED** [review]
